awscli/customizations/s3/s3handler.py

```diff
@@ -523,6 +525,80 @@ def _process_transfer(self, future):
             return CommandResult(1, 0)
 
 
+class S3TransferHandler(object):
+    def __init__(self, transfer_manager, cli_params, result_queue,
+                 result_recorder, result_processor):
+        """Backend for performing S3 transfers
+
+        :type transfer_manager: s3transfer.manager.TransferManager
+        :param transfer_manager: Transfer manager to use for transfers
+
+        :type cli_params: dict
+        :param cli_params: The parameters passed to the CLI command in the
+            form of a dictionary
+
+        :type result_queue: queue.Queue
+        :param result_queue: The queue to process results
+
+        :type result_recorder: ResultRecorder
+        :param result_recorder: Result recorder to get status of transfer
+
+        :type result_processor: ResultProcessor
+        :param result_processor: Result processor to process status updates
+            of transfer
+        """
+        self._transfer_manager = transfer_manager
+        self._result_queue = result_queue
```

Comment:
Interesting.  I was going down the path of abstracting away results so that code doesn't have to deal directly with `result_queue`, which is a bit of an implementation detail.  Something like:

```
with command_result_recorder as r:
    fileinfos = ...
    s3_transfer_handler.call(fileinfos, r)

# In s3 transfer handler
def call(..., reporter):
    try:
       # ...
    except Exception as e:
        reporter.report_result(ErrorResult(exception=e))
```

Sounds like what you're shooting for is to just abstract away the stop/start/final_result part of results and keep submission as is. I think that's an improvement over what's there now, but I'm concerned that separating out the lifecycle of the reporting (start/stop) from the mechanism for reporting (result_queue) could lead to subtle bugs, e.g submission to results queue after result thread is shutdown.